---
title: "RNASeqWorkflow : RNA-Seq workflow for one independent variable analysis"
author: "Author: Kuan-Hao Chao (ntueeb05howard@gmail.com)"
date: "Last update: `r format(Sys.time(), '%d %B, %Y')`"
abstract: >
  An RNA-Seq workflow for one independent variable analysis
output: 
  BiocStyle::html_document:
    highlight: pygments
    toc: true
    toc_depth: 3
    fig_caption: yes
bibliography: bibliography.bib
fontsize: 14pt
vignette: >
  %\VignetteIndexEntry{RNA-Seq workflow for one independent variable analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteDepends{}
  %\VignetteEncoding{UTF-8}
  %\usepackage[utf8]{inputenc}
  %\VignettePackage{RNASeqWorkflow}
---

```{r style, echo=FALSE, results="asis", message=FALSE}
knitr::opts_chunk$set(tidy = FALSE,
                      warning = FALSE,
                      message = FALSE)
```

# RNA-Seq workflow

## Overview

1. Each step is a function
2. Eahc function ==> R shell, background
3. 

Support paired-end reads only

## System Requirement
1. Python, 2to3
2. C++
3. 

## "RNASeqWorkFlowParam" S4 Object Creation

This is the **pre-step** of RNA-Seq workflow in this package. Before starting RNA-Seq analysis, you have to run the constructor function `RNASeqWorkFlowParam()` and create `RNASeqWorkFlowParam` S4 instance first. This instance will be used as input parameter in the following analysis function.

### `RNASeqWorkFlowParam` Slots Explanation
There are 11 slots in `RNASeqWorkFlowParam`.

1. `os.type` : The operating system type. Value is `linux` or `osx`. This package only support Linux and macOS (no Windows). If other operating system is detected, ERROR will be reported. 
2. `python.variable` : Python related variable. Value is a list of whether Python is available and Python version (`TRUE` or `FALSE`, `2` or `3`).
3. `python.2to3` : Availability of `2to3` command. Value is `TRUE` or `FALSE`.
4. `path.prefix` : Path prefix of *'gene_data/'*, *'RNASeqbin/'*, *'RNASeqresults/'*, *'Rscript/'* and *'Rscriptout/'* directories. It is recommended to create a new directory with out any file inside and all the following RNA-Seq results will be installed in it.
5. `input.path.prefix` : Path prefix of *'inputfiles/'* directory. User have to prepare an *'inputfile/'* directory with the following rules:
    * ***`genome.name`.fa*** : reference genome in FASTA file formation.
    * ***`genome.name`.gtf*** : gene annotation in GTF file formation.
    *  ***rawfastq.gz*/**: directory storing FASTQ files. 
    *  Support paired-end reads files only.
    *  Names of paired-end FASTQ files : *'`sample.pattern`1.fastq.gz'* and *'`sample.pattern`2.fastq.gz'*. `sample.pattern` must be distinct for each sample.
    * ***phenodata.csv*** : information about RNA-Seq experiment design.
        * First column : Distint ids for each sample. Value of each sample of this column must match `sample.pattern` in FASTQ files in *'rawfastq.gz/'*. Column names must be **ids**.
        * Second column : independent variable for the RNA-Seq experiment. Value of each sample of this column can only be parameter `case.group` and `control.group`. Column name is parameter `independent.variable`.
    * ***indices*/** : directory storing `HT2` indices files for hisat2 alignment tool. 
        * This directory is **optional**. `HT2` indices files corresponding to target reference genome can be installed at [HISAT2 official website](https://ccb.jhu.edu/software/hisat2/index.shtml). Providing `HT2` files can accelerate the subsequent steps.
        * If `HT2` index files are not provided, *'inputfiles/indices/'* directory should be deleted.
6. `genome.name` : Variable of genome name defined in this RNA-Seq workflow (ex. '`genome.name`.fa', '`genome.name`.gtf')  
7. `sample.pattern` : Regular expression of paired-end fastq.gz files under *'inputfiles/rawfastq.gz/'*. **IMPORTANT!!** expression shouldn't have `_[1,2].fastq.gz`  in end.
8. `independent.variable`: Independent variable for the biological experiment design of two-group RNA-Seq workflow.
9. `case.group` : Group name of the case group.
10. `control.group` : Group name of the control group.
11. `indices.optional` : logical value whether *'input_files/indices/'* is exit. Value is `TRUE` or `FALSE`

### `RNASeqWorkFlowParam` Constructor Checking Steps

1. Create a new directory for RNA-Seq analysis.
    * It is highly recommended to create a new directory without any files inside. 
The parameter `path.prefix` of `RNASeqWorkFlowParam()` constructor should be the absolute path of this new directory. All the RNA-Seq related files that generated in the following steps will be stored inside of this directory.

2. Create valid *'inputfiles/'* directory
    * You should create a file directory named *'inputfiles/'* with neccessary files inside. It should follow the rules metioned above.
    
3. Run constructor of `RNASeqWorkFlowParam` S4 object
    * This constructor will check following things before creating S4 objects
        * Operating system
        * Python version
        * 2to3 command  
        * Structure, contents and rules of *'inputfiles/'*
        * Validity of input parameters

### Example
```{r}
# exp <- RNASeqWorkFlowParam(path.prefix = NA, input.path.prefix = NA, 
# genome.name = NA, sample.pattern = NA, independent.variable = NA, 
# control.group = NA, experiment.group = NA)
```
In this example, `RNASeqWorkFlowParam` S4 object is store in `exp` for RNA-Seq analysis steps below. Any ERROR occured in checking steps will terminate the program.




## Environment Set up

This is the **first necessary step** of RNA-Seq workflow in this package. To set up the environment, run `RNASeqEnvironmentSet_CMD()` to execute process in background or run `RNASeqEnvironmentSet()` to execute process in R shell.

### Files Setup
1. Create Base Directories 
*'gene_data/'*, *'RNASeqbin/'*, *'RNASeqresults'*, *'Rscript'* and *'Rscriptout'* will be created under `path.prefix` directory.

2. Copy *'inputfiles/'*
Symbolic links will be created from files in *'inputfiles/'* to `path.prefix` directory. 

### Necessary Tools Installation
The operating system of your workstation will be detected. If the operating system is other than `Linux` and `macOS`, ERROR will be reported. `HISAT2`, `SAMtools`, `StringTie` and `Gffcompare` will be installed automatically.

* `HISAT2`
    * Based on your operating system, `hisat2-2.1.0-Linux_x86_64.zip` or `hisat2-2.1.0-OSX_x86_64.zip` zipped file will be installed.
    * Installed file will be unzipped and all binary files will be copied under *'RNASeqbin/'*
* *SAMtools*
    * `samtools-1.8.tar.bz2` source files will be installed. 
    * Installed files will be unzipped and compiled. If there is any ERROR occurred during compilation, program will be stopped. You should check whether the development environment is OK for SAMtools before rerunning `RNASeqEnvironmentSet_CMD` or `RNASeqEnvironmentSet()` to set up the environment again.
    * Maked samtools binary will be copied under *'RNASeqbin/'*.
* `StringTie`
    * Based on your operating system, `stringtie-1.3.4d.Linux_x86_64.tar.gz` or `stringtie-1.3.4d.Linux_x86_64` zipped file will be installed.
    * Installed file will be unzipped and all binary files will be copied under *'RNASeqbin/'*.
* `Gffcompare`
    * Based on your operating system, `gffcompare-0.10.4.Linux_x86_64.tar.gz` or `gffcompare-0.10.4.Linux_x86_64.tar.gz` zipped file will be installed.
    * Installed file will be unzipped and all binary files will be copied under *'RNASeqbin/'*.


### Export Path
*'RNASeqbin/'* will be added to R PATH so that these binaries can be found in R environment. In the last step of environment setting, `hisat2 --version`,`stringtie --version`,`gffcompare --version` and `samtools --version` commands will be checked in order to make sure the environment is correctly constructed.


### Example
1. Run in background
Result will be reported in *'Rscriptout/EnvironmentSet.Rout'*. Make sure the environment is successfully set up before running the subsequent steps.
```{r}
# RNASeqEnvironmentSet_CMD()
```

2. Run in R shell 
Result will be reported in R shell. Make sure the environment is successfully set up before running the subsequent steps.
```{r}
# RNASeqEnvironmentSet()
```


## `FASTQ` Files Quality Assessment

This is an **optional step** of RNA-Seq workflow in this package. However, it is strongly recommended to do it before processing raw reads. To assess the quality of raw reads in FASTQ files, run `RNASeqQualityAssessment_CMD()` to execute process in background or run `RNASeqQualityAssessment()` to execute process in R shell.

### "systemPipeR" Quality Assessment

In this process, [systemPipeR](http://bioconductor.org/packages/release/bioc/html/systemPipeR.html) package is used for quality assessment. Here are the following steps:

1. Check the number of times that user has run quality assessment process and create the corresponding files *'RNASeqresults/QAresults/QA{times}'*.
2. RNA-Seq environment set up. *'rnaseq/'* directory will be created by [systemPipeR](http://bioconductor.org/packages/release/bioc/html/systemPipeR.html) package.
3. Create *'data.list.txt'* file. 
4. Reading FASTQ files and create *'fastqReport.pdf'* as the report result of quality assessment
5. Remove *'rnaseq/'* directory.

### Example
1. Run in background
Result will be reported in *'Rscriptout/Quality_Assessment.Rout'*. Make sure quality assessment is successfully done before running the subsequent steps.
```{r}
# RNASeqQualityAssessment_CMD()
```

2. Run in R shell 
Result will be reported in R shell. Make sure quality assessment is successfully done before running the subsequent steps.
```{r}
# RNASeqQualityAssessment()
```

## `FASTQ` Files Quality Trimming

This is an **optional step** of RNA-Seq workflow in this package. However, it is strongly recommended to do it if the result of quality assessment is unsatisfied. To trim the raw reads in FASTQ files, run `RNASeqQualityTrimming_CMD()` to execute process in background or run `RNASeqQualityTrimming()` to execute process in R shell. After trimming process, it is strongly recommended to redo quality assessment step. Different quality assessment result will be stored in different directories with index of processing order.

### "ShortRead" Quality Trimming
In this process, [ShortRead](http://bioconductor.org/packages/release/bioc/html/ShortRead.html) package is used for quality trimming. Followings are trimming steps and method explanation.

1. `FASTQ` files with `sample.pattern` in *'gene_data/rawfastq.gz'* will be listed and be check. This trimming function only support paired-end FASTQ files.
2. The probability of error per each base pair is calculated by Phred Quality in FASTQ files and the cumulative probability of error of each base pair will also be calculated. The threshold cumulative probability `cum.error`, with default value `1`, is used to find the trimming point of paired end files.
3. `reads.length.limit`, with default value 36,  determines the minimum base pair length of reads. Length of reads less than `reads.length.limit` will be filtered out.
4. Untrimmed original `FASTQ` files are moved to *'gene_data/rawfastq.gz/originaluntrimmedfastq.gz'*, and trimmed reads are written to *'gene_data/rawfastq.gz/'*
5. It is strongly recommended to run `RNASeqQualityAssessment_CMD()` or `RNASeqQualityAssessment()` for quality assessment after quality trimming.

### Example
1. Run in background
Result will be reported in *'Rscriptout/QualityTrimming.Rout'*. Make sure quality trimming is successfully done before running the subsequent steps.
```{r}
# RNASeqQualityAssessment_CMD()
```

2. Run in R shell 
Result will be reported in R shell. Make sure quality assessment is successfully done before running the subsequent steps.
```{r}
# RNASeqQualityAssessment()
```

## Raw Reads Process (Alignment, Assembly, Gffcompare, Quantification)

This is a **second necessary step** of RNA-Seq workflow in this package. Programs installed in environment setting step, `HISAT2`, `SAMtools`, `StringTie` and `Gffcompare`, will be executed. To process raw reads FASTQ files, run `RNASeqRawReadProcess_CMD()` to execute process in background or run `RNASeqRawReadProcess()` to execute process in R shell. For more details about commands that executed during each reads processing step, please check the reported *'RNASeqresults/COMMAND.txt'*.

### `HISAT2` Index

In pre-step(`RNASeqWorkFlowParam` creation step), *'indices/'* directory is checked and the whether `HT2` indices files are provided is known. If not provided, following commands will be executed:

* Input: *`genome.name`.gtf*, *`genome.name`.fa*
* Output: *`genome.name`.ss*, *`genome.name`.exon*, *`genome.name`_tran.{number}.ht2*

1. `extract_splice_sites.py`, `extract_exons.py` execution
    * These two commands are executed to extract splice-site and exon information from gene annotation file.

2. `hisat2-build` index creation
    * This command build HISAT2 indices files with *`genome.name`.ss* and *`genome.name`.exon* created in the previous step. Be aware that index building step requires a larger amount of memory and longer time than the alignment step, and it might not be possible to run on some personal workstations. It is highly recommended to check the availibility of `HT2` indices files at [HISAT2 official website](https://ccb.jhu.edu/software/hisat2/index.shtml) for your target reference genome beforehands.

### `HISAT2` Alignment

* Input: *`genome.name`_tran.{number}.ht2*, *`sample.pattern`.fastq.gz*
* Output: *`sample.pattern`.sam*

1. `hisat2` command is executed on paired end FASTQ files. `SAM` files will be created.
    * `SAM` files are stored in *'gene_data/raw_bam/'*

### `SAMtools` `SAM` to `BAM`

* Input: *`sample.pattern`.sam*
* Output: *`sample.pattern`.bam*

1. `samtools` command is executed. Output files from `hisat2` `SAM` files will be converted to `BAM` files. 
    * `BAM` files are stored in *'gene_data/raw_sam/'*. 

### `StringTie` Assembly

* Input: *`genome.name`.gtf*, *`sample.pattern`.bam*
* Output: *`sample.pattern`.gtf*

1. `stringtie` command is executed. 
    * Assembler of RNA-Seq alignments into potential transcripts.
    * Output assembled `GTF` files which are from each `FASTQ` files are stored in *'gene_data/raw_gtf/'*

### `StringTie` Merge `GTF`

* Input: *`sample.pattern`.gtf*
* Output: *stringtiemerged.gtf*, *mergelist.txt*

1. Creating mergelist.txt
    * *gene_data/merged/mergelist.txt* is created for the merging step.
2. `stringtie` command is executed. 
    *  Transcript merger merges each *`sample.pattern`.gtf* into *stringtiemerged.gtf*
    *  Output files are all stored in *'gene_data/merged/'*

### `Gffcompare` 

* Input: *`genome.name`.gtf*, *stringtiemerged.gtf*
* Output: *merged.annotated.gtf*, *merged.loci*, *merged.stats*, *merged.stringtie_merged.gtf.refmap*, *merged.stringtie_merged.gtf.tmap*, *merged.tracking*

1. `gffcompare` command is executed. 
    * The comparison result of merged `GTF` file and reference annotation file is reported under *'merged/'* directory.

### Raw reads count table creation

Whether this step is executed depends on the availability of Python on your workstation. 

* Input: *samplelst.txt*
* Output: *gene_count_matrix.csv*, *transcript_count_matrix.csv*

1. Reads count table converter Python script is downloaded as `prepDE.py`
2. Python checking
    * If Python is not available, this step is skipped.
    * If Python2 is available, `prepDE.py` is executed.
    * If Python3 is available, `2to3` command will be checked.(Usally, if Python3 is installed, `2to3` command will be installed too.)
    * If Python3 is available but `2to3` command is unavailable, raw reads count generation step will be skipped.
    * If Python3 and `2to3` command is available, `prepDE.py` is converted to file that can be executed by Python2 and be executed.
3. raw reads count creation
    * Raw reads count is created and the results are stored in *'gene_data/reads_count_matrix/'*


### Example
1. Run in background
Result will be reported in *'Rscriptout/Raw_Read_Process.Rout'*. Make sure raw read process is successfully done before running the subsequent steps.
```{r}
# RNASeqRawReadProcess_CMD()
```

2. Run in R shell 
Result will be reported in R shell. Make sure raw read process is successfully done before running the subsequent steps.
```{r}
# RNASeqRawReadProcess()
```


## Differential Expressed Genes Analysis


This is the **third necessary step** of RNA-Seq workflow in this package. To do differential analysis(DA), run `RNASeqDifferentialAnalysis_CMD()` to execute process in background or run `RNASeqDifferentialAnalysis()` to execute process in R shell. In `RNASeqWorkflow` package, we provide four ways to do differential analysis. The reads sum of each row equals `0` in *`gene_data/reads_count_matrix/gene_count_matrix.csv`*, which is generated in the previous step, will be filtered out.

### "ballgown" Analysis

[Ballgown](https://www.bioconductor.org/packages/release/bioc/html/ballgown.html) is designed to facilitate flexible differential expression analysis of RNA-Seq data ([ballgown github](https://github.com/alyssafrazee/ballgown)). The default statistical model is a parametric F-test comparing nested linear models. Normalized row count in [ballgown](https://www.bioconductor.org/packages/release/bioc/html/ballgown.html) is represented as FPKM values. The following are [ballgown](https://www.bioconductor.org/packages/release/bioc/html/ballgown.html) analysis steps:

* Input: *'gene_data/ballgown/'*
* Output: *'RNASeq_results/ballgown_analysis/'*

1. `ballgown` object is created and written in *'RNASeqresults/ballgownanalysis/ballgownRobject/ballgown.rda'*. 
2. Row of extracted statistic results from created ballgown object with `pval` equals `0` or `qval` equals `0` will be filterd out. Log 2 fold change value will be calculated as column name `log2FC`.
3. The row sum of sample FPKM values equal '0' will be filter out. According to the *'gene_data/phenodata.csv'* file, normalized count will be seperated into case and control group with column name *`sample.pattern.FPKM`*.
4. A report `CSV` file combined with normalized FPKM count, average of case group, control group and statistic results will be stored as *'RNASeq_results/ballgown_analysis/ballgown_normalized_result.csv'*.
5. Differential expressed gene result *'RNASeq_results/ballgown_analysis/ballgown_normalized_DE_result.csv'* is reported. Default selecting threshold is **`pval < 0.05` and `log2FC > 1 | log2FC < 1`** 
6. Visualization. The following pictures will be plotted.
    * Pre differential analysis sample profile related plot
        * Frequency Plot:
            The frequency of normalized FPKM values will be drawn by using [rafalib](https://cran.r-project.org/web/packages/rafalib/index.html) package. X axis is `log2(FPKM)`; Y axis is `Frequency`. 
        * Box Plot & Violin Plot: 
            The distribution normalized FPKM values of each sample will be drawn by [ggplot2](https://ggplot2.tidyverse.org) package. X axis is `Samples`; Y axis is `Log2(FPKM+1)`. 
        * PCA Plot : 
            The principal components analysis(PCA) will be calculated on FPKM values by [FactoMineR](http://factominer.free.fr) and [factoextra](https://cran.r-project.org/web/packages/factoextra/index.html) packages. One dimension bar plot ([factoextra](https://cran.r-project.org/web/packages/factoextra/index.html)), two PCA plots ([factoextra](https://cran.r-project.org/web/packages/factoextra/index.html) and [graphics](https://stat.ethz.ch/R-manual/R-devel/library/graphics/html/00Index.html)) will be created.
        * Correlation Plot : 
            The correlation of each pair of samples will be calculated. One correlation dot plot ([corrplot](https://cran.r-project.org/web/packages/corrplot/vignettes/corrplot-intro.html)), one correlation bar plot ([PerformanceAnalytics](https://cran.r-project.org/web/packages/PerformanceAnalytics/PerformanceAnalytics.pdf)) and one correlation heat plot ([ggplot2](https://ggplot2.tidyverse.org)) will be created.
            
    * Differential analysis related plot
        * Volcano Plot : 
            The upregulated and downregulated genes will colored with red and green. X axis is `log2(Fold Change)`; Y axis is `log2(p value)`. It is drawn by [graphics](https://stat.ethz.ch/R-manual/R-devel/library/graphics/html/00Index.html) package. 
        * MA Plot : 
            The upregulated and downregulated genes will colored with red and green. X axis is `log2(Fold Change)`; Y axis is `log2(p value)`. It is drawn by [graphics](https://stat.ethz.ch/R-manual/R-devel/library/graphics/html/00Index.html) package.
        * PCA Plot : 
            The principal components analysis(PCA) of selected differential expressed genes will be calculated on their FPKM values by [FactoMineR](http://factominer.free.fr) and [factoextra](https://cran.r-project.org/web/packages/factoextra/index.html) packages. One dimension bar plot ([factoextra](https://cran.r-project.org/web/packages/factoextra/index.html)), two PCA plots ([factoextra](https://cran.r-project.org/web/packages/factoextra/index.html) and [graphics](https://stat.ethz.ch/R-manual/R-devel/library/graphics/html/00Index.html)) will be created.
        * Heatmap Plot : 
            Heatmap will be ploted by using [pheatmap](https://cran.r-project.org/web/packages/pheatmap/pheatmap.pdf) package.
            
### TPM & Student t-test Analysis

Normalized count is represented as TPM values. The statistic model used here is Student's t-test. The following are analysis steps:

* Input: *gene_data/ballgown/*
* Output: *RNASeq_results/TPM_analysis/*

1. Calculate TPM values from [ballgown](https://www.bioconductor.org/packages/release/bioc/html/ballgown.html) FPKM values and save with column name *`sample.pattern.FPKM`*.
2. Use Student's t-test to calculate p-value and saved as column name `pval`.
3. Calculate fold change values as dividing average TPM values of control group by average TPM values of case group (Average case TPM / Average control TPM) with column name `fc`. Log2 fold change is also calculated and stored as `log2FC`.

4. A report `CSV` file combined with normalized TPM count, average of case group, control group and statistic results will be stored as *'RNASeq_results/TPM_analysis/TPM_normalized_result.csv'*.
5. Differential expressed gene result *'RNASeq_results/TPM_analysis/TPM_normalized_DE_result.csv'* is reported. Default selecting threshold is **`pval < 0.05` and `log2FC > 1 | log2FC < 1`** 

6. Visualization. The following pictures will be plotted.
    * Pre differential analysis sample profile related plot
        * Frequency Plot:
            The frequency of normalized TPM values will be drawn by using [rafalib](https://cran.r-project.org/web/packages/rafalib/index.html) package. X axis is `log2(TPM)`; Y axis is `Frequency`. 
        * Box Plot & Violin Plot: 
            The distribution normalized TPM values of each sample will be drawn by [ggplot2](https://ggplot2.tidyverse.org) package. X axis is `Samples`; Y axis is `Log2(TPM+1)`. 
        * PCA Plot : 
            The principal components analysis(PCA) will be calculated on TPM valueby [FactoMineR](http://factominer.free.fr) and [factoextra](https://cran.r-project.org/web/packages/factoextra/index.html) packages. One dimension bar plot ([factoextra](https://cran.r-project.org/web/packages/factoextra/index.html)), two PCA plots ([factoextra](https://cran.r-project.org/web/packages/factoextra/index.html) and [graphics](https://stat.ethz.ch/R-manual/R-devel/library/graphics/html/00Index.html)) will be created.
        * Correlation Plot : 
            The correlation of each pair of samples will be calculated. One correlation dot plot ([corrplot](https://cran.r-project.org/web/packages/corrplot/vignettes/corrplot-intro.html)), one correlation bar plot ([PerformanceAnalytics](https://cran.r-project.org/web/packages/PerformanceAnalytics/PerformanceAnalytics.pdf)) and one correlation heat plot ([ggplot2](https://ggplot2.tidyverse.org)) will be created.
    * Differential analysis related plot

        * Volcano Plot : 
            The upregulated and downregulated genes will colored with red and green. X axis is `log2(Fold Change)`; Y axis is `log2(p value)`. It is drawn by [graphics](https://stat.ethz.ch/R-manual/R-devel/library/graphics/html/00Index.html) package. 
        * MA Plot : 
            The upregulated and downregulated genes will colored with red and green. X axis is `log2(Fold Change)`; Y axis is `log2(p value)`. It is drawn by [graphics](https://stat.ethz.ch/R-manual/R-devel/library/graphics/html/00Index.html) package.
        * PCA Plot : 
            The principal components analysis(PCA) of selected differential expressed genes will be calculated on their TPM values by [FactoMineR](http://factominer.free.fr) and [factoextra](https://cran.r-project.org/web/packages/factoextra/index.html) packages. One dimension bar plot ([factoextra](https://cran.r-project.org/web/packages/factoextra/index.html)), two PCA plots ([factoextra](https://cran.r-project.org/web/packages/factoextra/index.html) and [graphics](https://stat.ethz.ch/R-manual/R-devel/library/graphics/html/00Index.html)) will be created.
        * Heatmap Plot : 
            Heatmap will be ploted by using [pheatmap](https://cran.r-project.org/web/packages/pheatmap/pheatmap.pdf) package.
            
            
### "DESeq2" Analysis

[DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html) estimate variance-mean dependence in raw reads count ([DESeq2 document](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)). The statistic model for differential expression is using negative binomial distributionl. [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html) package takes sequence depth and RNA composition into consideration and use median of ratios normalization(MRN) method to count normalization ([related site](https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html)). The follwing are the analysis steps.

* Input: *gene_data/reads_count_matrix/gene_count_matrix.csv*
* Output: *RNASeq_results/ballgown_analysis/*

1. Create `DESeqDataSet` object from raw reads count of genes by running `DESeqDataSetFromMatrix()` function in DESeq2.
2. Run `DESeq2()` function to do differential gene analysis.
3. A report `CSV` file combined with normalized MRN count, average of case group, control group and statistic results will be stored as *'RNASeq_results/DESeq2_analysis/DESeq2_normalized_result.csv'*.
4. Differential expressed gene result *'RNASeq_results/DESeq2_analysis/DESeq2_normalized_DE_result.csv'* is reported. Default selecting threshold is **`pval < 0.05` and `log2FC > 1 | log2FC < 1`** 
5. Visualization. The following pictures will be plotted.
    * Pre differential analysis sample profile related plot
        * Frequency Plot:
            The frequency of normalized FPKM values will be drawn by using [rafalib](https://cran.r-project.org/web/packages/rafalib/index.html) package. X axis is `log2(MRN)`; Y axis is `Frequency`. 
        * Box Plot & Violin Plot: 
            The distribution normalized FPKM values of each sample will be drawn by [ggplot2](https://ggplot2.tidyverse.org) package. X axis is `Samples`; Y axis is `Log2(MRN+1)`. 
        * PCA Plot : 
            The principal components analysis(PCA) will be calculated on MRN values by [FactoMineR](http://factominer.free.fr) and [factoextra](https://cran.r-project.org/web/packages/factoextra/index.html) packages. One dimension bar plot ([factoextra](https://cran.r-project.org/web/packages/factoextra/index.html)), two PCA plots ([factoextra](https://cran.r-project.org/web/packages/factoextra/index.html) and [graphics](https://stat.ethz.ch/R-manual/R-devel/library/graphics/html/00Index.html)) will be created.
        * Correlation Plot : 
            The correlation of each pair of samples will be calculated. One correlation dot plot ([corrplot](https://cran.r-project.org/web/packages/corrplot/vignettes/corrplot-intro.html)), one correlation bar plot ([PerformanceAnalytics](https://cran.r-project.org/web/packages/PerformanceAnalytics/PerformanceAnalytics.pdf)) and one correlation heat plot ([ggplot2](https://ggplot2.tidyverse.org)) will be created.
            
    * Differential analysis related plot
        * Volcano Plot : 
            The upregulated and downregulated genes will colored with red and green. X axis is `log2(Fold Change)`; Y axis is `log2(p value)`. It is drawn by [graphics](https://stat.ethz.ch/R-manual/R-devel/library/graphics/html/00Index.html) package. 
        * MA Plot : 
            The upregulated and downregulated genes will colored with red and green. X axis is `log2(Fold Change)`; Y axis is `log2(p value)`. It is drawn by [graphics](https://stat.ethz.ch/R-manual/R-devel/library/graphics/html/00Index.html) package.
        * PCA Plot : 
            The principal components analysis(PCA) of selected differential expressed genes will be calculated on MRN values by [FactoMineR](http://factominer.free.fr) and [factoextra](https://cran.r-project.org/web/packages/factoextra/index.html) packages. One dimension bar plot ([factoextra](https://cran.r-project.org/web/packages/factoextra/index.html)), two PCA plots ([factoextra](https://cran.r-project.org/web/packages/factoextra/index.html) and [graphics](https://stat.ethz.ch/R-manual/R-devel/library/graphics/html/00Index.html)) will be created.
        * Heatmap Plot : 
            Heatmap will be ploted by using [pheatmap](https://cran.r-project.org/web/packages/pheatmap/pheatmap.pdf) package.
        * Dispersion Plot : 
            DESeq2 provide a useful function, `plotDispEsts()`, to plot the dispersion estimates. X axis is `mean of normalized counts`; Y axis is `dispersion`.

### "edgeR" Analysis

[edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html), a tool for differential expression analysis, implements a range of statistical method based on negative binomial distribution. [edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html) package normalizes for RNA composition by finding a set of scaling factors for the library sizes, which is computed by trimmed mean of M-values(TMM). Normalized count is represented as cpm with library size scaled by TMM ([related site](https://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf)). The following are analysis steps:

* Input: *gene_data/reads_count_matrix/gene_count_matrix.csv*
* Output: *RNASeq_results/edgeR_analysis/*

1. Create `DEGList` object from raw reads count by running `DGEList()` function in edgeR.
2. Normalize `DEGList` object with TMM by running `calcNormFactors()` function in edgeR.
3. Run `estimateCommonDisp()` to maximize the negative binomial conditional common likelihood and then run `estimateTagwiseDisp()` to estimate tagwise dispersion values by an empirical Bayes method.
4. Compute genewise exact tests and get statistic results by running `exactTest()`.
5. Get normalized count by running `cpm()` on TMM scaled `DEGList` object.
6. A report `CSV` file combined with normalized TMM count, average of case group, control group and statistic results will be stored as *'RNASeq_results/edgeR_analysis/edgeR_normalized_result.csv'*.
7. Differential expressed gene result *'RNASeq_results/edgeR_analysis/edgeR_normalized_DE_result.csv'* is reported. Default selecting threshold is **`pval < 0.05` and `log2FC > 1 | log2FC < 1`** 
8. Visualization. The following pictures will be plotted.

   * Pre differential analysis sample profile related plot
        * Frequency Plot:
            The frequency of normalized values will be drawn by using [rafalib](https://cran.r-project.org/web/packages/rafalib/index.html) package. X axis is `log2(MRN)`; Y axis is `Frequency`. 
        * Box Plot & Violin Plot: 
            The distribution normalized values of each sample will be drawn by [ggplot2](https://ggplot2.tidyverse.org) package. X axis is `Samples`; Y axis is `Log2(MRN+1)`. 
        * PCA Plot : 
            The principal components analysis(PCA) will be calculated by [FactoMineR](http://factominer.free.fr) and [factoextra](https://cran.r-project.org/web/packages/factoextra/index.html) packages. One dimension bar plot ([factoextra](https://cran.r-project.org/web/packages/factoextra/index.html)), two PCA plots ([factoextra](https://cran.r-project.org/web/packages/factoextra/index.html) and [graphics](https://stat.ethz.ch/R-manual/R-devel/library/graphics/html/00Index.html)) will be created.
        * MDS Plot : 
            The multidimensional scaling (MDS) visualization is provided by `plotMDS.DGEList()` in edgeR. Distance between dots in two dimensional scatterplot approximate the expression differences between the samples.
        * Correlation Plot : 
            The correlation of each pair of samples will be calculated. One correlation dot plot ([corrplot](https://cran.r-project.org/web/packages/corrplot/vignettes/corrplot-intro.html)), one correlation bar plot ([PerformanceAnalytics](https://cran.r-project.org/web/packages/PerformanceAnalytics/PerformanceAnalytics.pdf)) and one correlation heat plot ([ggplot2](https://ggplot2.tidyverse.org)) will be created.
        * MeanVar Plot (!!! Question ! preDE or DE ?): 
            Visualize the mean-variance relationship in `DEGList`. Plot is drawn by run `plotMeanVar()` in edgeR. 
        * plotBCV (!!! Question ! preDE or DE ?): 
            Plot the genewise biological coefficient of variance(BCV) against gene abundace.
            
    * Differential analysis related plot
        * Volcano Plot : 
            The upregulated and downregulated genes will colored with red and green. X axis is `log2(Fold Change)`; Y axis is `log2(p value)`. It is drawn by [graphics](https://stat.ethz.ch/R-manual/R-devel/library/graphics/html/00Index.html) package. 
        * MA Plot : 
            The upregulated and downregulated genes will colored with red and green. X axis is `log2(Fold Change)`; Y axis is `log2(p value)`. It is drawn by [graphics](https://stat.ethz.ch/R-manual/R-devel/library/graphics/html/00Index.html) package.
        * PCA Plot : 
            The principal components analysis(PCA) of selected differential expressed genes will be calculated on MRN values by [FactoMineR](http://factominer.free.fr) and [factoextra](https://cran.r-project.org/web/packages/factoextra/index.html) packages. One dimension bar plot ([factoextra](https://cran.r-project.org/web/packages/factoextra/index.html)), two PCA plots ([factoextra](https://cran.r-project.org/web/packages/factoextra/index.html) and [graphics](https://stat.ethz.ch/R-manual/R-devel/library/graphics/html/00Index.html)) will be created.
        * Heatmap Plot : 
            Heatmap will be ploted by using [pheatmap](https://cran.r-project.org/web/packages/pheatmap/pheatmap.pdf) package.
        * Dispersion Plot : 
            DESeq2 provide a useful function, `plotDispEsts()`, to plot the dispersion estimates. X axis is `mean of normalized counts`; Y axis is `dispersion`.





## Functional Analysis (Gene Ontology) and Pathway Analysis (Kyoto Encyclopedia of Genes and Genomes)

## Conlusion




genome.name Variable of genome name defined in this RNA-Seq workflow (ex. \code{genome.name}.fa, \code{genome.name}.gtf)


sample.pattern  Regular expression of paired-end fastq.gz files under 'input_files/raw_fastq.gz'. Expression not includes \code{_[1,2].fastq.gz}.

