---
title: "RNASeqWorkflow : RNA-Seq workflow for one independent variable analysis"
author: "Author: Kuan-Hao Chao (ntueeb05howard@gmail.com)"
date: "Last update: `r format(Sys.time(), '%d %B, %Y')`"
abstract: >
  An RNA-Seq workflow for one independent variable analysis
output: 
  BiocStyle::html_document:
    highlight: pygments
    toc: true
    toc_depth: 3
    fig_caption: yes
bibliography: bibliography.bib
fontsize: 14pt
vignette: >
  %\VignetteIndexEntry{RNA-Seq workflow for one independent variable analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteDepends{}
  %\VignetteEncoding{UTF-8}
  %\usepackage[utf8]{inputenc}
  %\VignettePackage{RNASeqWorkflow}
---

```{r style, echo=FALSE, results="asis", message=FALSE}
knitr::opts_chunk$set(tidy = FALSE,
                      warning = FALSE,
                      message = FALSE)
```

# RNA-Seq workflow

## Overview

1. Each step is a function
2. Eahc function ==> R shell, background
3. 

Support paired-end reads only

## System Requirement
1. Python, 2to3
2. C++
3. 

## "RNASeqWorkFlowParam" S4 Object Creation

This is the **pre-step** of RNA-Seq workflow in this package. Before starting RNA-Seq analysis, you have to run the constructor function `RNASeqWorkFlowParam()` and create `RNASeqWorkFlowParam` S4 instance first. This instance will be used as input parameter in the following analysis function.

### "RNASeqWorkFlowParam" Constructor Checking Steps

1. Create a new directory for RNA-Seq analysis.
    * It is highly recommended to create a new directory without any files inside. 
The parameter `path.prefix` of `RNASeqWorkFlowParam()` constructor should be the absolute path of this new directory. All the RNA-Seq related files that generated in the following steps will be stored inside of this directory.

2. Create valid 'input_files/' directory
    * You should create a file directory named 'input_files/' with the following rules:
        * ***`genome.name`.fa*** : reference genome in FASTA file formation.
        * ***`genome.name`.gtf*** : gene annotation in GTF file formation.
        *  ***rawfastq.gz*/**: directory storing FASTQ files. 
            *  Support paired-end reads files only.
            *  Names of paired-end FASTQ files : '`sample.pattern`_1.fastq.gz' and '`sample.pattern`_2.fastq.gz'. `sample.pattern` must be distinct for each sample.
        * ***phenodata.csv*** : information about RNA-Seq experiment design.
            * First column : Distint ids for each sample. Value of each sample of this column must match `sample.pattern` in FASTQ files in 'raw_fastq.gz/'. Column names must be 'ids'.
            * Second column : independent variable for the RNA-Seq experiment. Value of each sample of this column can only be parameter `control.group` and `experiment.group`. Column name is parameter `independent.variable`.
        * ***indices*/** : directory storing `HT2` indices files for hisat2 alignment tool.
            * This directory is optional. `HT2` indices files corresponding to target reference genome can be installed at [HISAT2 official website](https://ccb.jhu.edu/software/hisat2/index.shtml). Providing `HT2` files can accelerate the subsequent steps.
            * If `HT2` index files are not provided, 'indices/' directory should be deleted in 'input_files/' directory.
    
3. Run constructor of `RNASeqWorkFlowParam` S4 object
    * This constructor will check following things before creating S4 objects
        * Operating system
            * Detect your workstations operating system. This package support `Linux` and `macOS`. `Windows` is not supported.
        * Python version
            * Detect whether Python is available on your work station. 
        * Structure and contents of 'input_files/'
        * Validity of input parameters

### Example
```{r}
# exp <- RNASeqWorkFlowParam(path.prefix = NA, input.path.prefix = NA, 
# genome.name = NA, sample.pattern = NA, independent.variable = NA, 
# control.group = NA, experiment.group = NA)
```
In this example, `RNASeqWorkFlowParam` S4 object is store in `exp` for RNA-Seq analysis steps below.







## Environment Set up

This is the **first necessary step** of RNA-Seq workflow in this package. To set up the environment, run `RNASeqEnvironmentSet_CMD()` to execute process in background or run `RNASeqEnvironmentSet()` to execute process in R shell.

### Files Setup
1. Create Base Directories 
'gene_data/', 'RNASeq_bin/', 'RNASeq_results', 'Rscript' and 'Rscript_out' will be created under `path.prefix` directory.

2. Copy 'input_files/'
Symbolic links will be created from files in 'input_files/' to `path.prefix` directory. 

### Necessary Tools Installation
The operating system of your workstation will be detected. If the operating system is other than `Linux` and `macOS`, `ERROR` will be reported. `HISAT2`, `SAMtools`, `StringTie` and `Gffcompare` will be installed automatically.

* HISAT2
    * Based on your operating system, `hisat2-2.1.0-Linux_x86_64.zip` or `hisat2-2.1.0-OSX_x86_64.zip` zipped file will be installed.
    * Installed file will be unzipped and all binary files will be moved under 'RNASeq_bin/'

* SAMtools
    * `samtools-1.8.tar.bz2` source files will be installed. 
    * Installed files will be unzipped and compiled. If there is any ERROR occurred during compilation, program will be stopped. You should check the development environment is OK for SAMtools and rerun `RNASeqEnvironmentSet_CMD` or `RNASeqEnvironmentSet()` to set up the environment.
* StringTie
    * Based on your operating system, `stringtie-1.3.4d.Linux_x86_64.tar.gz` or `stringtie-1.3.4d.Linux_x86_64` zipped file will be installed.
    * Installed file will be unzipped and all binary files will be moved under 'RNASeq_bin/'
* Gffcompare
    * Based on your operating system, `gffcompare-0.10.4.Linux_x86_64.tar.gz` or `` zipped file will be installed.
    * Installed file will be unzipped and all binary files will be moved under 'RNASeq_bin/'


### Export Path
'RNASeq_bin/' will be added to R PATH so that these binaries can be found in R environment. In the last step of environment setting, `hisat2 --version`,`stringtie --version`,`gffcompare --version` and `samtools --version` commands will be checked in order to make sure the environment is correctly constructed.



### Example



## Fastq files Quality Assessment

This is an **optional step** of RNA-Seq workflow in this package. However, it is strongly recommended to do it before processing raw reads. To assess the quality of raw reads in FASTQ files, run `RNASeqQualityAssessment_CMD()` to execute process in background or run `RNASeqQualityAssessment()` to execute process in R shell.

### "systemPipeR" Quality Assessment

In this process, `systemPipeR` package is used for quality assessment. Here are the following steps:

1. Check the times that users have run quality assessment process and create the corresponding files 'RNASeq_results/QA_results/QA_{times}'.
2. RNA-Seq environment set up. 'rnaseq/' directory will be created by `systemPipeR` package.
3. Create 'data.list.txt' file. 
4. Reading FASTQ files and create 'fastqReport.pdf' as the report result of quality assessment
5. Remove 'rnaseq/' directory.

### Example


## Fastq files Quality Trimming

This is an **optional step** of RNA-Seq workflow in this package. However, it is strongly recommended to do it if the result of quality assessment is unsatisfied. To trim the raw reads in FASTQ files, run `RNASeqQualityTrimming_CMD()` to execute process in background or run `RNASeqQualityTrimming()` to execute process in R shell. After trimming process, it is strongly recommended to redo quality assessment step. Different quality assessment result will be stored in different directories with index of processing order.

### "ShortRead" Quality Trimming
In this process, `ShortRead` package is used for quality trimming. Followings are trimming steps and method explanation.

1. FASTQ files with `sample.pattern` in 'gene_data/raw_fastq.gz' will be listed and be check. This trimming function only support paired-end FASTQ files.
2. The probability of error per each base pair is calculated by Phred Quality in FASTQ files and the cumulative probability of error of each base pair will also be calculated. The threshold cumulative probability `cum.error`, with default value `1`, is used to find the trimming point of paired end files.
3. `reads.length.limit`, with default value 36,  determines the minimum base pair length of reads. Length of reads less than `reads.length.limit` will be trimmed.
4. Untrimmed FASTQ files are moved to 'gene_data/raw_fastq.gz/original_untrimmed_fastq.gz', and trimmed FASTQ files are written to 'gene_data/raw_fastq.gz/'
5. It is strongly recommended to run `RNASeqQualityAssessment_CMD()` or `RNASeqQualityAssessment()` for quality assessment after quality trimming.

### Example

## Raw Reads Process (Alignment, Assembly, Gffcompare, Quantification)

This is a **second necessary step** of RNA-Seq workflow in this package. Programs installed in environment setting step, `HISAT2`, `SAMtools`, `StringTie` and `Gffcompare`, will be executed. To process raw reads FASTQ files, run `RNASeqRawReadProcess_CMD()` to execute process in background or run `RNASeqRawReadProcess()` to execute process in R shell. For more details about commands that executed during each reads processing step, please check the reported `RNASeq_results/COMMAND.txt`.

### HISAT2 Index

In pre-step(`RNASeqWorkFlowParam` creation step), 'indices/' directory is checked and the whether `HT2` indices files are provided is known. If not provided, following commands will be executed:

* Input: *`genome.name`.gtf*, *`genome.name`.fa*
* Output: *`genome.name`.ss*, *`genome.name`.exon*, *`genome.name`_tran.{number}.ht2*

1. `extract_splice_sites.py`, `extract_exons.py` execution
    * These two commands are executed to extract splice-site and exon information from gene annotation file.

2. `hisat2-build` index creation
    * This command build HISAT2 indices files with *`genome.name`.ss* and *`genome.name`.exon* created in the previous step. Be aware that index building step requires a larger amount of memory and longer time than the alignment step, and it might not be possible to run on some personal workstations. It is highly recommended to check the availibility of `HT2` indices files at [HISAT2 official website](https://ccb.jhu.edu/software/hisat2/index.shtml) for your target reference genome beforehands.

### HISAT2 Alignment

* Input: *`genome.name`_tran.{number}.ht2*, *`sample.pattern`.fastq.gz*
* Output: *`sample.pattern`.sam*

1. `hisat2` command is executed on paired end FASTQ files. `SAM` files will be created.
    * `SAM` files are stored in 'gene_data/raw_bam/'

### SAMtools `SAM` to `BAM`

* Input: *`sample.pattern`.sam*
* Output: *`sample.pattern`.bam*

1. `samtools` command is executed. Output files from `hisat2` `SAM` files will be converted to `BAM` files. 
    * `BAM` files are stored in 'gene_data/raw_sam/'. 

### StringTie Assembly

* Input: *`genome.name`.gtf*, *`sample.pattern`.bam*
* Output: *`sample.pattern`.gtf*

1. `stringtie` command is executed. 
    * Assembler of RNA-Seq alignments into potential transcripts.
    * Output assembled `GTF` files which are from each `FASTQ` files are stored in 'gene_data/raw_gtf/'

### StringTie Merge `GTF`

* Input: *`sample.pattern`.gtf*
* Output: *stringtiemerged.gtf*, *mergelist.txt*

1. Creating *mergelist.txt*
    * *mergelist.txt* is created for the merging step.
2. `stringtie` command is executed. 
    *  Transcript merger merges each *`sample.pattern`.gtf* into *stringtiemerged.gtf*
    *  Output files are all stored in 'gene_data/merged/'

### Gffcompare 

* Input: *`genome.name`.gtf*, *stringtiemerged.gtf*
* Output: *merged.annotated.gtf*, *merged.loci*, *merged.stats*, *merged.stringtie_merged.gtf.refmap*, *merged.stringtie_merged.gtf.tmap*, *merged.tracking*

1. `gffcompare` command is executed. 
    * The comparison result of merged `GTF` file and reference annotation file is reported under 'merged/' directory.

### Raw reads count table creation

Whether this step is executed depends on the availability of Python on your workstation. 

* Input: *samplelst.txt*
* Output: *gene_count_matrix.csv*, *transcript_count_matrix.csv*

1. Reads count table converter Python script is downloaded as `prepDE.py`
2. Python checking
    * If Python is not available, this step is skipped.
    * If Python2 is available, `prepDE.py` is executed.
    * If Python3 is available, `2to3` command is checked.(Usally, if Python3 is installed, `2to3` command will be installed too.)
    * If Python3 is available but `2to3` command is unavailable, raw reads count generation step will be skipped.
3. `prepDE.py` raw reads count creation
    * Raw reads count is executed and the results are stored in 'gene_data/reads_count_matrix/'

### 

### Example


## Raw Reads Differential Expressed Genes Analysis (ballgown, TPM & t-test, DESeq2 and edgeR)

## Functional Analysis (Gene Ontology) and Pathway Analysis (Kyoto Encyclopedia of Genes and Genomes)

## Conlusion




genome.name Variable of genome name defined in this RNA-Seq workflow (ex. \code{genome.name}.fa, \code{genome.name}.gtf)


sample.pattern  Regular expression of paired-end fastq.gz files under 'input_files/raw_fastq.gz'. Expression not includes \code{_[1,2].fastq.gz}.
